src/libsync/propagatefiles.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) by Piotr Mrowczynski <piotr@owncloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "owncloudpropagator.h"
+
+namespace OCC {
+
+
+qint64 PropagateFiles::committedDiskSpace() const
+{
+    qint64 needed = 0;
+    foreach (PropagatorJob* job, _subJobs) {
+        needed += job->committedDiskSpace();
+    }
+    return needed;
+}
+
+void PropagateFiles::append(const SyncFileItemPtr &item)
+{
+    _totalItems++;
+    if(item->_size <= propagator()->smallFileSize()){
+        _syncDBItems.append(item);
+    } else {
+        _syncDataItems.append(item);
+    }
+}
+
+bool PropagateFiles::scheduleNextJob()
+{
+    if (_state == Finished) {
+        return false;
+    }
+
+    if (_state == NotYetStarted) {
+        _state = Running;
+
+        if (isEmpty()) {
+            finalize();
+            return true;
+        }
+        _subJobs.reserve(_totalItems);
+    }
+
+    if (_state == Running) {
+        // This will ensure that all other jobs in the earlier PropagateDirectory are finished, so we can start with data transfers
+        if (propagator()->runningAtRootJob() != 1){
+            return false;
+        }
+    }
+
+    // cache the value of first unfinished subjob
+    int i = _firstUnfinishedSubJob;
+    int subJobsCount = _subJobs.count();
+    while (i < subJobsCount && _subJobs.at(i)->_state == Finished) {
+      _firstUnfinishedSubJob = ++i;
+    }
+    for (int i = _firstUnfinishedSubJob; i < subJobsCount; ++i) {
+        if (_subJobs.at(i)->_state == Finished) {
+            continue;
+        }
+
+        if (possiblyRunNextJob(_subJobs.at(i))) {
+            return true;
+        }
+
+        Q_ASSERT(_subJobs.at(i)->_state == Running);
+    }
+
+    return scheduleNextItem();
+}
+
+bool PropagateFiles::scheduleNewJob(QVector<SyncFileItemPtr> &syncJobs){
+    // This function is used to schedule new job and lazily create job from sync items
+    Q_ASSERT(!syncJobs.isEmpty());
+    const SyncFileItemPtr &item = syncJobs.takeFirst();
```

Comment:
We are planning to use only 5.8 dont we? Or should I still keep compatibility with Qt4?